java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/processfunction/SlidingWindowKeyedProcessFunction.java

```diff
@@ -0,0 +1,190 @@
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Map;
+
+/**
+ * A KeyedProcessFunction that aggregate sliding windows with different sizes. The ProcessFunction
+ * should be applied after the input has been aggregated with a tumble window with size equals to
+ * the step size of sliding window.
+ *
+ * <p>One native Flink sliding window operator can only aggregate under one sliding window size. If
+ * we need aggregation under different window sizes, we need to apply multiple native Flink window
+ * operators with different sizes and join the results together. Each window operator keeps the rows
+ * for its window size. This will result in a lots of duplicates row in the state backend. For
+ * example, if we want to do aggregation under 1 hour and 2 hours, both sliding window will keep the
+ * rows in the last one hour, which are duplicated. With this process function, we only keep the
+ * rows for the maximum window size so that we can avoid duplicated rows in state backend.
+ *
+ * <p>The ProcessFunction assumes that: 1. rows of each key are ordered by the row time 2. row time
+ * attribute of rows with the same key are all distinct. The assumptions hold true after applying
+ * the tumbling window aggregation to the input with window size that is same as the step size of
+ * the {@link SlidingWindowKeyedProcessFunction} to be applied.
+ */
+public class SlidingWindowKeyedProcessFunction extends KeyedProcessFunction<Row, Row, Row> {
+
+    private final AggregationFieldsDescriptor aggregationFieldsDescriptor;
+    private final TypeSerializer<Row> rowTypeSerializer;
+    private final String rowTimeFieldName;
+    private final long stepSizeMs;
+    private MultiWindowSizeState state;
+
+    public SlidingWindowKeyedProcessFunction(
+            AggregationFieldsDescriptor aggregationFieldsDescriptor,
+            TypeSerializer<Row> rowTypeSerializer,
+            String rowTimeFieldName,
+            long stepSizeMs) {
+        this.aggregationFieldsDescriptor = aggregationFieldsDescriptor;
+        this.rowTypeSerializer = rowTypeSerializer;
+        this.rowTimeFieldName = rowTimeFieldName;
+        this.stepSizeMs = stepSizeMs;
+    }
+
+    @Override
+    public void open(Configuration parameters) {
+        state =
+                MultiWindowSizeState.buildMultiWindowSizeState(
+                        getRuntimeContext(), rowTypeSerializer);
+    }
+
+    @Override
+    public void processElement(
+            Row row, KeyedProcessFunction<Row, Row, Row>.Context ctx, Collector<Row> out)
+            throws Exception {
+        final long rowTime = ((Instant) row.getFieldAs(rowTimeFieldName)).toEpochMilli();
+        long triggerTime = rowTime;
+        while (triggerTime <= rowTime + aggregationFieldsDescriptor.getMaxWindowSizeMs()) {
+            ctx.timerService().registerEventTimeTimer(triggerTime);
```

Comment:
Since we already know that the inputs are ordered by time, would it be simpler and more performant to compute the results every time we receive an event without using timer?

Reply:
We have to compute the result using timers because even after the tumbling window operation, it doesn't guarantee that there is a row at every step, and we need to perform aggregation at every step.